vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,527 @@
-                    tags.extend(instance.get('tags', []))
+                # Get the most recent value that isn't negative
+                valid_values = [v for v in result.value if v >= 0]
+                if not valid_values:
+                    self.log.debug(
+                        "Skipping metric %s because the value returned by vCenter"
+                        " is negative (i.e. the metric is not yet available).",
+                        ensure_unicode(metric_name),
+                    )
+                    continue
+                value = valid_values[-1]
+                if metric_name in PERCENT_METRICS:
+                    # Convert the percentage to a float.
+                    value /= 100
+
+                tags = []
+                if should_collect_per_instance_values(metric_name, resource_type):
+                    instance_tag_key = get_mapped_instance_tag(metric_name)
+                    instance_tag_value = result.id.instance or 'none'
+                    tags.append('{}:{}'.format(instance_tag_key, instance_tag_value))
+
+                if resource_type in HISTORICAL_RESOURCES:
+                    # Tags are attached to the metrics
+                    tags.extend(mor_props['tags'])
+                    hostname = None
+                else:
+                    # Tags are (mostly) submitted as external host tags.
+                    hostname = ensure_unicode(mor_props.get('hostname'))
+                    if self.excluded_host_tags:
+                        tags.extend([t for t in mor_props['tags'] if t.split(":", 1)[0] in self.excluded_host_tags])
 
-                    # vsphere "rates" should be submitted as gauges (rate is
-                    # precomputed).
-                    self.gauge("vsphere.{}".format(ensure_unicode(metric_name)), value, hostname=hostname, tags=tags)
+                tags.extend(self.base_tags)
 
-        # ## <TEST-INSTRUMENTATION>
-        custom_tags = instance.get('tags', []) + ['instance:{}'.format(i_key)]
-        self.histogram('datadog.agent.vsphere.metric_colection.time', t.total(), tags=custom_tags)
-        # ## </TEST-INSTRUMENTATION>
+                # vsphere "rates" should be submitted as gauges (rate is
+                # precomputed).
+                self.gauge(ensure_unicode(metric_name), value, hostname=hostname, tags=tags)
 
-    def collect_metrics(self, instance):
+    def query_metrics_wrapper(self, query_specs):
+        """Just an instrumentation wrapper around the VSphereAPI.query_metrics method
+        Warning: called in threads
         """
-        Calls asynchronously _collect_metrics_async on all MORs, as the
-        job queue is processed the Aggregator will receive the metrics.
+        t0 = Timer()
+        metrics_values = self.api.query_metrics(query_specs)
+        self.histogram('datadog.vsphere.query_metrics.time', t0.total(), tags=self.base_tags, raw=True)
+        return metrics_values
+
+    def collect_metrics(self, thread_pool):
```

Comment:
Updated, let me know what you think

Reply:
(minor) I would pass directly the `results` to `submit_metrics_callback` instead of the `future`. That will help keep/isolate async code in `collect_metrics_async`.

Reply:
Done